components/salsa-2022/src/interned.rs

```diff
@@ -83,23 +83,28 @@ where
             self.reset_at,
         );
 
+        // Optimisation to only get read lock on the map if the data has already
+        // been interned.
         if let Some(id) = self.key_map.get(&data) {
             return *id;
         }
 
-        loop {
-            let next_id = self.counter.fetch_add(1);
-            let next_id = Id::from_id(crate::id::Id::from_u32(next_id));
-            match self.value_map.entry(next_id) {
-                // If we already have an entry with this id...
-                dashmap::mapref::entry::Entry::Occupied(_) => continue,
-
-                // Otherwise...
-                dashmap::mapref::entry::Entry::Vacant(entry) => {
-                    self.key_map.insert(data.clone(), next_id);
-                    entry.insert(Box::new(data));
-                    return next_id;
-                }
+        match self.key_map.entry(data.clone()) {
+            // Data has already been interned, just use the existing ID
+            dashmap::mapref::entry::Entry::Occupied(entry) => *entry.get(),
+            // Insert the new value at the next available id
+            // This holds a lock on the entry so other can't attempt to insert
+            // at the same time.
+            dashmap::mapref::entry::Entry::Vacant(entry) => {
+                let next_id = self.counter.fetch_add(1);
+                let next_id = Id::from_id(crate::id::Id::from_u32(next_id));
+                let old_value = self.value_map.insert(next_id, Box::new(data));
```

Comment:
My concern is that in `delete_index` we hold the locks in the reverse order:

https://github.com/salsa-rs/salsa/blob/61e3bdfc81463e7bf1ea9fa6d3896e3e29b2a7a4/components/salsa-2022/src/interned.rs#L161-L166

As the comment says:

https://github.com/salsa-rs/salsa/blob/61e3bdfc81463e7bf1ea9fa6d3896e3e29b2a7a4/components/salsa-2022/src/interned.rs#L35

If there were a call to `delete_index` occurring in parallel, it could deadlock.

That said, I think `delete_index` might be dead code? I wanted at some point to experiment with removing interned values, or having an LRU-like system, but I don't think we're using it at all now. We could delete it and address the question later.

Reply:
Well spotted, I read that comment but didn't quite internalise what it meant. We do use delete_index to clean up tracked structs, so I've implemented what I hope is a proper fix.